File: 019. Introducción a Spring WebFlux/spring-webflux/src/main/java/com/example/controller/BookFunctionalController.java
package com.example.controller;

import com.example.model.Book;
import com.example.service.BookService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.MediaType;
import org.springframework.web.reactive.function.BodyInserter;
import org.springframework.web.reactive.function.BodyInserters;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.RouterFunctions;
import org.springframework.web.reactive.function.server.ServerResponse;

import javax.print.attribute.standard.Media;
import java.net.URI;

@Configuration
public class BookFunctionalController {

    @Autowired
    private BookService bookService;

    @Bean
    public RouterFunction<ServerResponse> findAllBooks(){
        return RouterFunctions.route().GET(
                "/functional/books",
                request -> ServerResponse.ok().body(bookService.findAll(), Book.class)
        ).build();
    }

    @Bean
    public RouterFunction<ServerResponse> findById(){
        return RouterFunctions.route().GET(
                "/functional/books/{id}",
                request -> ServerResponse.ok().body(
                        bookService.findById(Long.valueOf(request.pathVariable("id"))),
                        Book.class
                )
        ).build();
    }

    @Bean
    public RouterFunction<ServerResponse> createBook(){
        return RouterFunctions.route().POST(
                "/functional/books",
                request -> request.bodyToMono(Book.class)
                        .flatMap(bookService::create)
                        .flatMap(book ->
                                ServerResponse.created(URI.create("/functional/books/" + book.getId()))
                                        .contentType(MediaType.APPLICATION_JSON)
                                        .body(BodyInserters.fromValue(book))
                        )
        ).build();
    }

}

File: 020. Introducción a Spring WebFlux con Kotlin/spring-webflux-kotlin/src/main/kotlin/com/example/controller/BookFunctionalController.kt
package com.example.controller

import com.example.service.BookHandler
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.MediaType.APPLICATION_JSON
import org.springframework.web.reactive.function.server.router

@Configuration
class BookFunctionalController {

    @Bean
    fun bookRouter(handler: BookHandler) = router {
        ("/api/books" and accept(APPLICATION_JSON)).nest {
            GET("", handler::findAll)
            GET("/{id}", handler::findById)
            POST("", handler::create)
            PUT("", handler::update)
            DELETE("/{id}", handler::findById)
            DELETE("", handler::deleteAll)
        }
    }
}